<h1>Chapter 3. Going beyond the Basics: Detecting Features in Images</h1>

In [10]:
import tensorflow as tf 
data = tf.keras.datasets.fashion_mnist
import numpy as np

In [11]:
(training_images, training_labels), (test_images, test_labels) = data.load_data()
print('train image shape:', training_images.shape[:])
print('test image shape:', test_images.shape[:])


train image shape: (60000, 28, 28)
test image shape: (10000, 28, 28)


In [12]:
training_images = training_images.reshape(60000, 28, 28, 1)
training_images = training_images/255
test_images = test_images.reshape(10000, 28, 28, 1)
test_images = test_images/255

Because **Conv2D** layers are designed for multicolor channel processing (RGB as example). Therefore, the original 28 x 28 image need to be reshaped as (28x28x1). 

In [13]:
model = tf.keras.models.Sequential([
    tf.keras.layers.Conv2D(64, (3,3), activation = 'relu', input_shape = (28, 28, 1)), 
    tf.keras.layers.MaxPooling2D(2, 2), 
    tf.keras.layers.Conv2D(64, (3, 3), activation = 'relu'), 
    tf.keras.layers.MaxPooling2D(2, 2), 
    tf.keras.layers.Flatten(), 
    tf.keras.layers.Dense(128, activation = tf.nn.relu), 
    tf.keras.layers.Dense(10, activation = tf.nn.softmax)
])

c:\Users\DELL\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\convolutional\base_conv.py:107: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [14]:
model.compile(optimizer = 'adam', 
              loss = 'sparse_categorical_crossentropy', 
              metrics = ['accuracy'])

In [7]:
model.fit(training_images, training_labels, epochs = 20)

Epoch 1/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 31s 15ms/step - accuracy: 0.7801 - loss: 0.6099
Epoch 2/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 27s 14ms/step - accuracy: 0.8892 - loss: 0.3036
Epoch 3/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 28s 15ms/step - accuracy: 0.9067 - loss: 0.2509
Epoch 4/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 27s 14ms/step - accuracy: 0.9182 - loss: 0.2152
Epoch 5/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 25s 13ms/step - accuracy: 0.9283 - loss: 0.1900
Epoch 6/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 12ms/step - accuracy: 0.9375 - loss: 0.1664
Epoch 7/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 12ms/step - accuracy: 0.9479 - loss: 0.1395
Epoch 8/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 12ms/step - accuracy: 0.9513 - loss: 0.1280
Epoch 9/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 22s 12ms/step - accuracy: 0.9579 - loss: 0.1117
Epoch 10/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 23s 12ms/step - accuracy: 0.9642 - loss: 0.0959
Epoch 11/20
1875/1875 ━━━━━━━━━━━━━━━━━━━━ 24s 13ms/step - accuracy: 0.9685 - loss: 0.0830
Epoch 12

In [8]:
model.summary()

Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d (Conv2D)                 │ (None, 26, 26, 64)     │           640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d (MaxPooling2D)    │ (None, 13, 13, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_1 (Conv2D)               │ (None, 11, 11, 64)     │        36,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_1 (MaxPooling2D)  │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten (Flatten)               │ (None, 1600)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense (Dense)                   │ (None, 128)            │       204,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 731,360 (2.79 MB)

 Trainable params: 243,786 (952.29 KB)

 Non-trainable params: 0 (0.00 B)

 Optimizer params: 487,574 (1.86 MB)

<h2>Parameter calculation</h2>

<li><strong>Conv2d</strong>: Input shape: (28 x 28 x1) -> Output shape: (26 x 26 x 64). This layer applies 64 conv filters into input image, resulting 64 dimensional-reduced output images (26 x 26). <strong>General formular</strong>: n x n kernel leads to the output image of [x-(n-1)]^2. <strong>Parameter calculation</strong>: 64 kernel, each of which (3 x 3) is applied to the input image, plus the bias term added into each convolution: 64 x (3 x 3 + 1) = 640  </li>
<li><strong>Max pooling</strong>: conduct (2 x 2) max-pooling, therefore reduce the total image dimension by a quarter</li>
<li><strong>Conv2d</strong>: Input shape (13 x 13 x 54) -> Output shape: (11 x 11 x 54). This layer applies 64 conv filters into the existing 64 convolved image of previous layer. <strong>Parameter calculation</strong>: 64 x (64 x 9) + 64 = 36928</li>
<li></strong>Max pooling</strong>: Same as above</li>
<li><strong>Flatten</strong>: Break all the input array into 1-D array.</li>



In [9]:
result = model.evaluate(test_images, test_labels)

313/313 ━━━━━━━━━━━━━━━━━━━━ 2s 5ms/step - accuracy: 0.9069 - loss: 0.5337
